### PR TITLE
Bug 1094759 - new bootstrap

### PR DESCRIPTION
**Pull request by alivedise**

### CODE REVIEW

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=a089c2f0794268c1e278e6abc50c9753db749faa

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=dc912ed619cf76f2ae879098523142351c9c4234

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=d27d4bdc42548fa7491298df38d961a59a372e68

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=1f73fa99d7ba2a4743dd347d11abf4b2f6e79794

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=b67667d7af0da7012b3a60866cda3b183e8cc9a9

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=cb7b4d05113d4970b0ea67efc9496fa1ae0aff48

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=154508b522e01f8f4347cddf2b23a71ffa99499d

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=e70ee361b9305f1103e4daace4a885842f7272af

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=ba94055580f7dc430232adf61cec99066c7a6d32

File: apps/system/js/statusbar.js [hunk 1]
```diff
@@ -14,597 +14,567 @@
...
+      window.addEventListener('homescreenopening', this);
+      window.addEventListener('homescreenopened', this);
+      window.addEventListener('stackchanged', this);
+
+      // Listen to updates dialog
+      window.addEventListener('updatepromptshown', this);
+      window.addEventListener('updateprompthidden', this);
+
+      // We need to preventDefault on mouse events until
+      // https://bugzilla.mozilla.org/show_bug.cgi?id=1005815 lands
```
Comment (mikehenrty) [inline apps/system/js/statusbar.js]: Can we remove this now? https://bugzilla.mozilla.org/show_bug.cgi?id=1005815#c38

File: apps/system/js/statusbar.js [hunk 2]
```diff
@@ -14,597 +14,567 @@
...
+        this.topPanel.addEventListener(name, this.panelHandler.bind(this));
+      }, this);
+
+      this.statusbarIcons.addEventListener('wheel', this);
+
+      LazyLoader.load(['js/utility_tray.js']).then(function() {
+        this.utilityTray = UtilityTray;
+        UtilityTray.init();
+      }.bind(this)).catch((err) => {
+        console.error(err);
```
Comment (mikehenrty) [inline apps/system/js/statusbar.js]: nit: this error msg should mention something about attempting to load the utility tray.

File: apps/system/js/statusbar.js [hunk 3]
```diff
@@ -14,597 +14,567 @@
...
-
-    // Only calculate the search input width when the chrome is minimized
-    // Bug 1118025 for more info
-    if (appChrome && appChrome.isMaximized()) {
-      this._updateIconVisibility();
-      return;
-    }
+      this.element.classList.toggle('maximized', app.isHomescreen ||
+        !!(app.appChrome && app.appChrome.isMaximized()) ||
+           app.isLockscreen);
```
Comment (mikehenrty) [inline apps/system/js/statusbar.js]: We removed the check for `app.isAttentionWindow` that we need for bug 1159885. Also, where does the `app.isLockscreen` part come from?

File: apps/system/js/statusbar.js [hunk 4]
```diff
@@ -14,597 +14,567 @@
...
+      if (Service.query('isFtuRunning')) {
+        return;
+      }
 
-    // Do not forward events is utility-tray is active
-    if (UtilityTray.active) {
-      return;
-    }
+      // Do not forward events is utility-tray is active
+      if (window.UtilityTray && window.UtilityTray.active) {
```
Comment (mikehenrty) [inline apps/system/js/statusbar.js]: nit: this should probably be `this.utilityTray`

File: apps/system/js/statusbar.js (hunk 1 above)
Comment (alivedise) [inline apps/system/js/statusbar.js]: Maybe, but out of scope of this patch.

File: apps/system/js/statusbar.js (hunk 2 above)
Comment (alivedise) [inline apps/system/js/statusbar.js]: I think the err will mention it..but as you wish.

File: apps/system/js/statusbar.js (hunk 3 above)
Comment (alivedise) [inline apps/system/js/statusbar.js]: It's to correct the internal variable here
https://github.com/mozilla-b2g/gaia/blob/master/apps/system/js/statusbar.js#L370

Use lockscreen-appopen/lockscreen-appclosing to check if top most window is lockscreen is wrong. There might be a secure-window opened on top of lockscreen when this._inLockScreenMode = true

The app.isLockScreen is true fo LockscreenWindow instance see the lockscreen_window.js in the patch.

File: apps/system/js/statusbar.js (hunk 4 above)
Comment (alivedise) [inline apps/system/js/statusbar.js]: Good, though we still have window.UtilityTray. Fixed

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=8191f7a8c61a4f8066698d4b44664113893265a3

File: apps/keyboard/test/marionette/dismiss_test.js [hunk 5]
```diff
@@ -30,9 +30,11 @@ marionette('Dimiss the keyboard', function() {
   }
 
   setup(function() {
+    system = client.loader.getAppClass('system');
+    system.waitForFullyLoaded();
     actions = client.loader.getActions();
     systemInputMgmt = client.loader.getAppClass('system', 'input_management');
-    system = client.loader.getAppClass('system');
+
```
Comment (etiennesegonzac) [inline apps/keyboard/test/marionette/dismiss_test.js]: I think we have trailing spaces here that the linter should've caught.

File: apps/system/js/browser_mixin.js [hunk 6]
```diff
@@ -346,5 +350,7 @@
     }
   };
 
-  AppWindow.addMixin(BrowserMixin);
-}(this));
+  if (exports.AppWindow) {
+    exports.AppWindow.addMixin(BrowserMixin);
+  }
```
Comment (etiennesegonzac) [inline apps/system/js/browser_mixin.js]: testing issue ?

File: apps/system/test/unit/edge_swipe_detector_test.js [hunk 7]
```diff
@@ -1030,17 +1024,35 @@ suite('system/EdgeSwipeDetector >', function() {
...
-  suite('handleEvent: lifecycle events', function() {
+  suite('Test hierarchy changed event', function() {
+    test('Hierarchy top most ui is appWindowManager', function() {
+      MockService.mockQueryWith('getTopMostUI', {
+        name: 'AppWindowManager'
+      });
+      MockService.mockQueryWith('getTopMostWindow', { isHomescreen: false});
+      window.dispatchEvent(new CustomEvent('hierarchychanged'));
+      assert.isTrue(subject.lifecycleEnabled);
+    });
```
Comment (etiennesegonzac) [inline apps/system/test/unit/edge_swipe_detector_test.js]: need to cover the homescreen case here too

**Comment by etiennesegonzac** [comment]
`scheduler_test.js` is not displayed by github, but I'd like to add a test making sure we actually resolve the promises returned.

File: apps/system/test/marionette/statusbar_home_button_test.js [hunk 8]
```diff
@@ -52,6 +52,8 @@ marionette('Closing statusbar via home button >', function() {
            .release()
            .perform();
 
+    client.helper.wait(2000);
```
Comment (etiennesegonzac) [inline apps/system/test/marionette/statusbar_home_button_test.js]: We need to figure this one out. We wouldn't let a random wait land as part of a small patch we can't let it slip here.

File: apps/system/test/marionette/text_selection_test.js [hunk 9]
```diff
@@ -390,14 +394,17 @@ marionette('Text selection >', function() {
...
+        // XXX: Use system.turnScreenOff();
         clientWithLockscreen.switchToFrame();
         clientWithLockscreen.executeScript(function() {
-          window.wrappedJSObject.ScreenManager.turnScreenOff(true, 'powerkey');
+          window.wrappedJSObject.Service.request(
+            'turnScreenOff', true, 'powerkey');
         });
         clientWithLockscreen.helper.wait(500);
         // turn on screen
+        // XXX: Use system.turnScreenOff();
```
Comment (etiennesegonzac) [inline apps/system/test/marionette/text_selection_test.js]: s/turnScreenOff/turnScreenOn/

File: apps/system/test/unit/app_test.js [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+  var fakeCoreStartDeferred;
+
+  var Deferred = function() {
+    this.promise = new Promise(function(resolve, reject) {
+      this.resolve = resolve;
+      this.reject = reject;
+    }.bind(this));
+
+    return this;
+  };
```
Comment (etiennesegonzac) [inline apps/system/test/unit/app_test.js]: this is reapeated a few time, might be worth it to put it in a file

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=6f8018ba89e13d6ccf42b09f370530af00308ef9

File: apps/system/test/unit/app_test.js (hunk 10 above)
Comment (alivedise) [inline apps/system/test/unit/app_test.js]: done

File: apps/system/test/marionette/statusbar_home_button_test.js (hunk 8 above)
Comment (alivedise) [inline apps/system/test/marionette/statusbar_home_button_test.js]: removed, should be fixed by something...

File: apps/system/test/unit/edge_swipe_detector_test.js (hunk 7 above)
Comment (alivedise) [inline apps/system/test/unit/edge_swipe_detector_test.js]: done

File: apps/system/js/browser_mixin.js (hunk 6 above)
Comment (alivedise) [inline apps/system/js/browser_mixin.js]: Loading order issue.

File: apps/keyboard/test/marionette/dismiss_test.js (hunk 5 above)
Comment (alivedise) [inline apps/keyboard/test/marionette/dismiss_test.js]: fixed

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=c483f5904730c02c26650b44581e78d546ee5d10

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=2f2b1bfef52a14e92c91908ee5a07220e9bfdaae

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=b7aa34b9ad8f8f02394fc16608e75577669e9c77

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=a85552d1943770bdf194d52b10d2250d9f557709

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=f25b43949649d290639de7d5493cd58eb87d07fc

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=dddc3051ac1fb7ae7d838d95ee0c69a4de9fadc6

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=06230476e01a202aacb6699c8fbb791f1c43f5c3

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=e038f31a73b464afbeebadeb707895d89bebd5e9

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=f9e3d2c0f7afe3059fab0970e4f2628e13b06aa2

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=35cbb65cb24d09387d8a7d78abcbe3957474b6b2

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=b4aaa429d321159cfd97a97a5527f139e369da62

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=6889e703b5145e5db0fe9f4fb3ffe59d55964ec0

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=089856c2cc987b1498f143c8c0afb6223b11ce2a

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=33795a5e4c582a5f0d797e325bb8b54f29f64290

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=f968d6efe1ec92b2e9d4f0018003258fe7ac5233

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=9ed52a815de135b8ec3c139bfb84bae7c59be5f5

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=f0467e43dfdd7d07cc03dfb371e3bcb084f560a0

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=723e8aa18ea23991767e8391c631625c161f236a

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=82c6d02aa9844f6d916e5313fb48958681e25d61

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=821e5d2cf76bfd89707077ec429da7d3b4cc220a

**Comment by autolander** [comment]
The pull request could not be applied to the integration branch. Please try again after current integration is complete. You may need to rebase your branch against the target branch.

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=7137faaa6476e5a57bbd85de54251b4cc20c1c17

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=5637e9ce97fdd6175cb91085f6f0fdec22bfc508

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=24e25d94d9da2c32ec77ae2cc9dae800f13c0430

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=595a2fca79ce95737cda8db7733eafe51c483c58

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=94bc990b9048ac4d5199727ed09a374adf7133b0

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=481db507e20c3c6940056093de114281b8bf69c4

**Pull request merged**